src/aml/parser.rs

```diff
@@ -0,0 +1,512 @@
+use super::stream::AmlStream;
+use super::value::{AmlValue, FieldFlags, RegionSpace};
+use super::{opcodes, AmlError};
+use alloc::{string::ToString, String};
+use bit_field::BitField;
+use core::str;
+use {Acpi, AcpiHandler};
+
+/// This is used internally by the parser. Often, we're only interested in the end offset, so we
+/// know when to stop parsing the current explicit-length structure. However, various constants
+/// (e.g. the size of fields) are encoded as PkgLengths, and so sometimes we want to access the raw
+/// data as well.
+struct PkgLength {
+    pub raw_length: u32,
+    pub end_offset: u32,
+}
+
+pub(crate) struct AmlParser<'s, 'a, 'h, H>
+where
+    'h: 'a,
+    H: AcpiHandler + 'h,
+{
+    acpi: &'a mut Acpi<'h, H>,
+    scope: String,
+    stream: AmlStream<'s>,
+}
+
+/// This macro takes a parser and one or more parsing functions and tries to parse the next part of
+/// the stream with each one. If a parsing function fails, it rolls back the stream and tries the
+/// next one. If none of the functions can parse the next part of the stream, we error on the
+/// unexpected byte.
+// TODO: rewrite to not use try_parse, and just `match` the errors manually. Might be much cleaner
+macro_rules! try_parse {
+    ($parser: expr, $($function: path),+) => {
+        if false {
+            unreachable!();
```

Comment:
What's this `if false` for?

Reply:
Yeah, this is a bit of a hacky way of writing this macro (I think I left a TODO to redo it). Basically, each function's snippet needed to begin with `else if ...` so this branch will never pass and be optimised out, but satisfies the parser

Reply:
Um... why??

Reply:
Looking at this again, it wasn't a brilliant way of writing it. How's the new version?